### PR TITLE
stop using BytePos for computing spans in librustc_parse/parser/mod.rs

### LINKED ISSUE #68783: Reduce use of `span.with_lo(span.lo() + BytePos(1))` in `librustc_parse/parser/mod.rs`
`span.with_lo(span.lo() + BytePos(1))` may cause an ICE if span points strange character so we should use `next_point` instead. But `next_point` isn't actually equivalent with `span.with_lo(span.lo() + BytePos(1))` so it breaks some spans in current diagnostics. We should find out the cause and reduce use of them.

See context: https://github.com/rust-lang/rust/pull/68735#discussion_r373872254

### PR DESCRIPTION
**Pull request by dwrensha**
Computing spans using logic such as `self.token.span.lo() + BytePos(1)` can cause internal compiler errors like #68730 when non-ascii characters are given as input. 

#68735 partially addressed this problem, but only for one case. Moreover, its usage of `next_point()` does not actually align with what `bump_with()` expects. For example, given the token `>>=`, we should pass the span consisting of the final two characters `>=`, but `next_point()` advances the span beyond the end of the `=`. 

This pull request instead computes the start of the new span by doing `start_point(self.token.span).hi()`. This matches `self.token.span.lo() + BytePos(1)` in the common case where the characters are ascii, and it gracefully handles multibyte characters.

Fixes #68783.

### CODE REVIEW

**Comment by dwrensha** [comment]
(Repushed after an `./x.py fmt`.)

**Comment by dwrensha** [comment]
The diff in the test stderr is:

```

---- [ui] ui/parser/issue-68730.rs stdout ----
diff of stderr:

33         |        ^
34
35      error: expected one of `>`, `const`, identifier, or lifetime, found `<`
-         --> $DIR/issue-68730.rs:5:11
+         --> $DIR/issue-68730.rs:5:10
37         |
38      LL | enumem˂˂
-          |         ^ expected one of `>`, `const`, identifier, or lifetime
+          |        ^ expected one of `>`, `const`, identifier, or lifetime
40
41      error: aborting due to 5 previous errors
42

```

File: src/librustc_parse/parser/mod.rs [hunk 1]
```diff
@@ -707,16 +712,16 @@ impl<'a> Parser<'a> {
...
             token::BinOpEq(token::Shr) => {
-                let span = self.token.span.with_lo(self.token.span.lo() + BytePos(1));
-                Some(self.bump_with(token::Ge, span))
+                let start_point = self.sess.source_map().start_point(self.token.span);
+                Some(self.bump_with(token::Ge, self.token.span.with_lo(start_point.hi())))
             }
             token::Ge => {
-                let span = self.token.span.with_lo(self.token.span.lo() + BytePos(1));
-                Some(self.bump_with(token::Eq, span))
+                let start_point = self.sess.source_map().start_point(self.token.span);
```
Comment (estebank) [inline src/librustc_parse/parser/mod.rs]: If this is such a common thing then we might want to have a new method in `SourceMap` for it.

**Pull request merged**